src/Features/Core/Portable/ExtractInterface/AbstractExtractInterfaceService.cs

```diff
@@ -110,6 +113,9 @@ public ExtractInterfaceResult ExtractInterfaceFromAnalyzedType(ExtractInterfaceT
                 ? string.Empty
                 : refactoringResult.TypeToExtractFrom.ContainingNamespace.ToDisplayString();
 
+            // GetExtractInterfaceOptions will potentially show a UI dialog, and may be 
+            // expected to mantain the same calling thread. After this call, 
+            // everything else can be done in a background thread
             var extractInterfaceOptions = GetExtractInterfaceOptions(
```

Comment:
hmm.. so I looked http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ExtractInterface/VisualStudioExtractInterfaceOptionsService.cs,17

I believe the VSExtractInterfaceOptionsService itself should switch to UI thread if it needs to. by using IThreadingContext (http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/GoToDefinition/AbstractGoToSymbolService.cs,38)

not the caller trying to maintain UI thread to call it.

I would change all Synchronous/WaitAndGetResult to real async and change actual call to extract interface to use Run(http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Experimentation/VisualStudioExperimentationService.cs,28) to call async method from UI thread.

tagging @sharwell